repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala

```diff
@@ -0,0 +1,585 @@
+
+  var keystoneV2Config: KeystoneV2Config = _
+
+  override def init(filterConfig: FilterConfig): Unit = {
+    configurationFile = new FilterConfigHelper(filterConfig).getFilterConfig(DEFAULT_CONFIG)
+    logger.info(s"Initializing Keystone V2 Filter using config $configurationFile")
+    val xsdURL: URL = getClass.getResource("/META-INF/schema/config/keystone-v2.xsd")
+    configurationService.subscribeTo(
+      SYSTEM_MODEL_CONFIG,
+      getClass.getResource("/META-INF/schema/system-model/system-model.xsd"),
+      SystemModelConfigListener,
+      classOf[SystemModel]
+    )
+    configurationService.subscribeTo(
+      filterConfig.getFilterName,
+      configurationFile,
+      xsdURL,
+      KeystoneV2ConfigListener,
+      classOf[KeystoneV2Config]
+    )
+  }
+
+  override def destroy(): Unit = {
+    configurationService.unsubscribeFrom(configurationFile, KeystoneV2ConfigListener)
+    configurationService.unsubscribeFrom(SYSTEM_MODEL_CONFIG, SystemModelConfigListener)
+  }
+
+  override def doFilter(servletRequest: ServletRequest, servletResponse: ServletResponse, chain: FilterChain): Unit = {
+    /**
+     * STATIC REFERENCE TO CONFIG
+     */
+    val config = keystoneV2Config
+
+    /**
+     * DECLARE COMMON VALUES
+     */
+    lazy val request = MutableHttpServletRequest.wrap(servletRequest.asInstanceOf[HttpServletRequest])
+    // Not using the mutable wrapper because it doesn't work properly at the moment, and
+    // we don't need to modify the response from further down the chain
+    lazy val response = servletResponse.asInstanceOf[HttpServletResponse]
+    lazy val traceId = Option(request.getHeader(CommonHttpHeader.TRACE_GUID.toString)).filter(_ => sendTraceHeader)
+    lazy val requestHandler = new KeystoneRequestHandler(config.getIdentityService.getUri, akkaServiceClient, traceId)
+
+    /**
+     * BEGIN PROCESSING
+     */
+    if (!isInitialized) {
+      logger.error("Keystone v2 filter has not yet initialized")
+      servletResponse.asInstanceOf[HttpServletResponse].sendError(SC_INTERNAL_SERVER_ERROR)
+    } else {
+      logger.trace("Keystone v2 filter processing request...")
+
+      val keystoneAuthenticateHeader = s"Keystone uri=${config.getIdentityService.getUri}"
+
+      val filterResult =
+        if (isWhitelisted(request.getRequestURI)) {
+          Pass
+        } else {
+          val processingResult =
+            getAuthToken flatMap { authToken =>
+              validateToken(getAdminToken, authToken) flatMap { validToken =>
+                addTokenHeaders(validToken)
+                authorizeTenant(tenantFromUri, validToken) flatMap { _ =>
+                  lazy val endpoints = getEndpoints(getAdminToken, authToken, validToken) // Prevents making call if its not needed
+                  addCatalogHeader(endpoints) flatMap { _ =>
+                    authorizeEndpoints(validToken, endpoints) flatMap { _ =>
```

Comment:
I don't remember the exact syntax, but you do have this method parameter endpoints setup correctly so that it isn't getting evaluated here, but when it might be needed in the method body itself right?